src/framed_read.rs

```diff
@@ -125,24 +125,25 @@ where
 
         let mut buf = [0u8; INITIAL_CAPACITY];
 
-        loop {
-            let n = ready!(Pin::new(&mut this.inner).poll_read(cx, &mut buf))?;
-            this.buffer.extend_from_slice(&buf[..n]);
-
-            match this.inner.decode(&mut this.buffer)? {
-                Some(item) => return Poll::Ready(Some(Ok(item))),
-                None => {
-                    if this.buffer.is_empty() {
-                        return Poll::Ready(None);
-                    } else if n == 0 {
-                        return Poll::Ready(Some(Err(io::Error::new(
-                            io::ErrorKind::UnexpectedEof,
-                            "bytes remaining in stream",
-                        )
-                        .into())));
-                    }
-                }
-            }
+        let n = ready!(Pin::new(&mut this.inner).poll_read(cx, &mut buf))?;
+        this.buffer.extend_from_slice(&buf[..n]);
+
+        let ended = n == 0;
+
+        match this.inner.decode(&mut this.buffer)? {
+            Some(item) => Poll::Ready(Some(Ok(item))),
+            None if ended => Poll::Ready(if this.buffer.is_empty() {
+                None
+            } else {
+                // this is the end of the input but there are bytes left
+                // maybe do something like tokio's `decode_eof` here instead?
+                Some(Err(io::Error::new(
+                    io::ErrorKind::UnexpectedEof,
+                    "bytes remaining in stream",
+                )
+                .into()))
+            }),
+            _ => Poll::Pending,
```

Comment:
It looks to me like the `Stream` here would return `Poll::Pending` without a way to wake the context's `Waker`. If that's the case I should probably cover this in the tests

Reply:
hmm yeah, i had to do a few `wake_by_ref()` calls inside my decoder impl to prevent it from hanging. I don't really know what the right way is to use Wakers, just added them in places until it worked.

I guess if we got some data, but didn't get an item from it yet, it makes sense to immediately do `wake_by_ref()` so the future is polled again and we can pull more data from the underlying AsyncRead?

Reply:
Actually in that case wrapping it in a `loop {}` thing becomes useful again, so we can immediately poll the AsyncRead if we need more data, and the ASyncRead can do the waking once it's ready